File: gitlfs-common/src/main/java/ru/bozaro/gitlfs/common/io/InputStreamValidator.java
package ru.bozaro.gitlfs.common.io;

import org.jetbrains.annotations.NotNull;
import ru.bozaro.gitlfs.common.data.Meta;

import javax.xml.bind.DatatypeConverter;
import java.io.IOException;
import java.io.InputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Wrapper for validating hash and size of uploading object.
 *
 * @author Artem V. Navrotskiy
 */
public class InputStreamValidator extends InputStream {
  @NotNull
  private final MessageDigest digest;
  @NotNull
  private final InputStream stream;
  @NotNull
  private final Meta meta;

  private boolean eof;
  private long totalSize;

  public InputStreamValidator(@NotNull InputStream stream, @NotNull Meta meta) throws IOException {
    try {
      this.digest = MessageDigest.getInstance("SHA-256");
    } catch (NoSuchAlgorithmException e) {
      throw new IOException(e);
    }
    this.stream = stream;
    this.meta = meta;
    this.eof = false;
    this.totalSize = 0;
  }

  @Override
  public int read() throws IOException {
    if (eof) {
      return -1;
    }
    final int data = stream.read();
    if (data >= 0) {
      digest.update((byte) data);
      checkSize(1);
    } else {
      checkSize(-1);
    }
    return data;
  }

  private void checkSize(int size) throws IOException {
    if (size > 0) {
      totalSize += size;
    }
    if ((meta.getSize() > 0 && totalSize > meta.getSize())) {
      throw new IOException("Input stream too big");
    }
    if (size < 0) {
      eof = true;
      if ((meta.getSize() >= 0) && (totalSize != meta.getSize())) {
        throw new IOException("Unexpected end of stream");
      }
      final String hash = DatatypeConverter.printHexBinary(digest.digest());
      if (!meta.getOid().equals(hash)) {
        throw new IOException("Invalid stream hash");
      }
    }
  }

  @Override
  public int read(@NotNull byte[] buffer, int off, int len) throws IOException {
    if (eof) {
      return -1;
    }
    final int size = stream.read(buffer, off, len);
    if (size > 0) {
      digest.update(buffer, off, size);
    }
    checkSize(size);
    return size;
  }

  @Override
  public void close() throws IOException {
    stream.close();
  }
}

File: gitlfs-common/src/main/kotlin/ru/bozaro/gitlfs/common/io/InputStreamValidator.kt
package ru.bozaro.gitlfs.common.io

import ru.bozaro.gitlfs.common.data.Meta
import java.io.IOException
import java.io.InputStream
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

/**
 * Wrapper for validating hash and size of uploading object.
 *
 * @author Artem V. Navrotskiy
 */
class InputStreamValidator(stream: InputStream, meta: Meta) : InputStream() {
    private var digest: MessageDigest? = null

    private val stream: InputStream

    private val meta: Meta
    private var eof: Boolean
    private var totalSize: Long

    @Throws(IOException::class)
    override fun read(): Int {
        if (eof) {
            return -1
        }
        val data = stream.read()
        if (data >= 0) {
            digest!!.update(data.toByte())
            checkSize(1)
        } else {
            checkSize(-1)
        }
        return data
    }

    @Throws(IOException::class)
    private fun checkSize(size: Int) {
        if (size > 0) {
            totalSize += size.toLong()
        }
        if (meta.size in 1 until totalSize) {
            throw IOException("Input stream too big")
        }
        if (size < 0) {
            eof = true
            if (meta.size >= 0 && totalSize != meta.size) {
                throw IOException("Unexpected end of stream")
            }
            val hash = toHexString(digest!!.digest())
            if (meta.oid != hash) {
                throw IOException("Invalid stream hash")
            }
        }
    }

    @Throws(IOException::class)
    override fun read(buffer: ByteArray, off: Int, len: Int): Int {
        if (eof) {
            return -1
        }
        val size = stream.read(buffer, off, len)
        if (size > 0) {
            digest!!.update(buffer, off, size)
        }
        checkSize(size)
        return size
    }

    @Throws(IOException::class)
    override fun close() {
        stream.close()
    }

    companion object {
        private val hexDigits = "0123456789abcdef".toCharArray()
        private fun toHexString(bytes: ByteArray): String {
            val sb = StringBuilder(2 * bytes.size)
            for (b in bytes) {
                sb.append(hexDigits[b.toInt() shr 4 and 0xf]).append(hexDigits[b.toInt() and 0xf])
            }
            return sb.toString()
        }
    }

    init {
        try {
            digest = MessageDigest.getInstance("SHA-256")
        } catch (e: NoSuchAlgorithmException) {
            throw IOException(e)
        }
        this.stream = stream
        this.meta = meta
        eof = false
        totalSize = 0
    }
}
